### PR TITLE
[sparse] semi-structured sparse refactor

### PR DESCRIPTION
**Pull request by jcaip**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #119704
* __->__ #117302

Summary:

This PR is a refactor of semi-structured sparsity support.

**deprecation**:

Before `torch.sparse.to_sparse_semi_structured` had a kwarg param
`transposed=False`, which has been removed. This kwarg was unused and
now thros a deprecation warning.

Namely, I've taken the subclassing implementation that xFormers has
created and brought it over to PyTorch, as part of our plan to upstream
runtime 2:4 sparsity.

I've also copied over all the op support that Daniel implemenented that
did not depend on the fast sparsification routines, into
`_sparse_semi_structured_ops.py`

With this subclass, all of our internal tests pass, as well as those in
xFormers.

The main change is that we now define a base subclass,
`SparseSemiStructuredTensor` that is inherited from for each of the
specific backends.

We also now can arbitrarily override the sparse dispatch table with
`_load_dispatch_table()`, idea being this is still general enough
where users don't need to modify pytorch source code to get their model
working.

This also adds in padding support and stores alg_id and fuse_transpose
as flags on the tensor, instead of hardcoding them.

There still remains two components in xFormers that will need to be
ported over eventually:
- the autograd functions  (`Sparsify24`, `Sparsify24_like`)
- fast sparsification routines that they rely on

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

### CODE REVIEW

**Comment by alexsamardzic** [comment]
I like the changes, the code is cleaner indeed.  Are we sure that having only `mm()` method (that actually implements `addmm`) in cuSPARSELt/CUTLASS specific C++ code is enough?  If so, I could simplify CUTLASS-based code, but I'm asking as CUTLASS version also provides fusion with activation functions, and could provide more fusions in the future, for example with de-quantization.

Nit: maybe we should avoid  having`24` in method names (as for `int32` it's actually `12`, and there may be more sparsity patterns supported in the future)?

**Comment by jcaip** [comment]
> I like the changes, the code is cleaner indeed. Are we sure that having only `mm()` method (that actually implements `addmm`) in cuSPARSELt/CUTLASS specific C++ code is enough? If so, I could simplify CUTLASS-based code, but I'm asking as CUTLASS version also provides fusion with activation functions, and could provide more fusions in the future, for example with de-quantization.

I'll update this to addmm and the signature to be more general, so we won't break bc. I think it should be enough, since we can always just change the dispatch table. Ideally support for CUTLASS epilogue fusions would be supported with this API. 

I guess this method is really `mm` in name only, but with support for all the backend specific / fusion options. Do you think we need to add more ops? 

> Nit: maybe we should avoid having`24` in method names (as for `int32` it's actually `12`, and there may be more sparsity patterns supported in the future)?

Yes, I'll clean that up.

**Comment by alexsamardzic** [comment]
> I guess this method is really `mm` in name only, but with support for all the backend specific / fusion options. Do you think we need to add more ops?
> 

I'll defer that to @cpuhrsch.  On the other side, something to keep in mind is that we'd probably want to have functionality of these methods (`torch._cslt_sparse_mm` and `torch._sparse_semi_structured_linear`) also supported by Inductor - definitely through lowering to corresponding ATen operators, but also hopefully through providing options to auto-tuner to select the fastest among several lowerings possible.

**Review by alexsamardzic — APPROVED** [review]
LGTM.

**Review by HDCharles — APPROVED** [review]
looks good to me

File: torch/sparse/semi_structured.py [hunk 1]
```diff
@@ -100,387 +119,189 @@ def __new__(
...
-        )
+    @classmethod
+    def from_dense(cls, original_tensor : torch.Tensor) -> "SparseSemiStructuredTensor":
+        raise NotImplementedError
 
-        return sparse_semi_structured_to_dense_cutlass(
-            self.sparse_tensor_cutlass,
-            self.meta_tensor_cutlass,
-        )
+    def _mm(
```
Comment (HDCharles) [inline torch/sparse/semi_structured.py:297]: nit: i tend to prefer these to be class methods rather than attached to a specific instance of the class

**Pull request closed without merging**